Next.js quickstart proposal (#444)
clerkinc/clerk-docs: docs/quickstarts/nextjs.mdx

description: Learn how to use Clerk to quickly and easily add secure authentication and user management to your Next.js application. Clerk works seamlessly on both client side and server side components.
---

> # Add authentication and user management to your Next.js app with Clerk

Reviewer: If this is the h1, we need to update the title in the frontmatter to match.

## You will learn how to

* Install `@clerk/nextjs`
* Set up your environment keys
* Wrap your Next.js app in `<ClerkProvider />`
* Limit access to authenticated users
* Embed the `<UserButton />`

**Estimated time** 1 cup of coffee

> ### Prerequisites 

Reviewer: We can add a little more context to these bullet points. So for example, our social connection pages have this on every page:
## Before you start
- You need to create a Clerk Application in your [Clerk Dashboard](https://dashboard.clerk.com/). For more information, check out our [Set up your application guide](https://clerk.com/docs/quickstarts/setup-clerk).
- You need to have a <Insert Social connection> developer account. To create one, visit the [<Insert Social connection> console](#).

E.g. see [Google social connection](https://clerk.com/docs/authentication/social-connections/google) page.

> * [Set up Clerk](/docs/quickstarts/setup-clerk)

Author: What's the reasoning behind its deletion?

> * [Set up Clerk](/docs/quickstarts/setup-clerk)

Author: It is not getting deleted

* [A Next.js project](https://nextjs.org/docs/getting-started/installation)

#### Get the finished project

<Tabs items={["App Router version", "Page Router version"]}>
  <Tab>
    [`github.com/clerkinc/clerk-next-app-router-starter`](https://github.com/clerkinc/clerk-next-app-router-starter)
  </Tab>
  <Tab>
    [`github.com/clerkinc/clerk-nextjs-starter`](https://github.com/clerkinc/clerk-nextjs-starter)
  </Tab>
</Tabs>

<Steps>

### Install `@clerk/nextjs`

> Clerk's Next.js SDK gives you access to prebuilt components and hooks and [helpers](/docs/references/nextjs/overview) for Next.js API routes, server-side rendering, and Middleware. Install it by running the following command in your terminal:

Reviewer: yes. and should have reference links, so
[components](https://clerk.com/docs/components/overview)
and
[hooks](/docs/references/nextjs/overview#client-side-helpers)

</InjectKeys>

This sets your public and secret keys.

### Wrap your app in `<ClerkProvider />`

The [`<ClerkProvider />`](/docs/components/clerk-provider) component provides active session and user context to Clerk's hooks and other components. Import it into your app by adding `import { ClerkProvider } from '@clerk/nextjs'` at the top of your file. 

<Tabs items={["App Router", "Pages Router"]}>
  <Tab>
    
    <Callout>
      The `app/layout.tsx` file might be in your `src` folder if it isn't in your root folder.
    </Callout>

    For this project, it makes sense to wrap the `<html />` element with `<ClerkProvider />`. This makes the active session and user context accessible anywhere within the app.

    ```tsx filename="/src/app/layout.tsx" {1,9,13}
    import { ClerkProvider } from '@clerk/nextjs'

    export default function RootLayout({
      children,
    }: {
      children: React.ReactNode
    }) {
      return (
        <ClerkProvider>
          <html lang="en">

>             <body className={inter.className}>{children}</body>

Reviewer 2: ```suggestion
            <body>{children}</body>

```

### Set environment keys

> In your Next.js project's root folder, you may have an `.env.local` file alongside `package.json` and other configuration files. If you don't see it, create it.

Reviewer 3: "If you don't see it, create it" -- Love this cause I always find myself lost (sometimes we also have `.env.local.example`)

Author: I walked into this wall myself lol

Add the following code to your `.env.local` file to set your public and secret keys.

> **Pro tip!** If you are signed into your [Clerk Dashboard](https://dashboard.clerk.com/), your keys should become visible by clicking on the eye icon.

Reviewer: I agree. Usually if there is a "tip", it's good opportunity to use a callout with type="info".

> **Pro tip!** If you are signed into your [Clerk Dashboard](https://dashboard.clerk.com/), your keys should become visible by clicking on the eye icon.

Author: I chose not to use a callout for this because it felt unworthy of being a callout. This page previously looked like a patchwork of callouts, and there can be a tendency in docs to "make everything a call out." Call outs are, by their nature, disruptive to reading. I didn't think this was worth disrupting their flow. Perhaps marrying it into the preceding content is a better idea?

</InjectKeys>

### Wrap your app in `<ClerkProvider />`

> The [`<ClerkProvider />`][clerk-provider] component provides active session and user context to Clerk's hooks and other components. For this project, it makes sense to wrap the `<html/>` element with `<ClerkProvider />`. This makes the active session and user context accessible anywhere within the app.

Reviewer 3: We had specifically decided to move away from early education around opting into dynamic rendering when wrapping with the `ClerkProvider` right?

> The [`<ClerkProvider />`][clerk-provider] component provides active session and user context to Clerk's hooks and other components. For this project, it makes sense to wrap the `<html/>` element with `<ClerkProvider />`. This makes the active session and user context accessible anywhere within the app.

Author: Getting rid of it as per styleguide

> The [`<ClerkProvider />`][clerk-provider] component provides active session and user context to Clerk's hooks and other components. For this project, it makes sense to wrap the `<html/>` element with `<ClerkProvider />`. This makes the active session and user context accessible anywhere within the app.

Author: Thanks for pointing this out! I've slid this information into the tab

<Tabs items={["App Router version", "Page Router version"]}>
  <Tab>
    
    <Callout>
      The `app/layout.tsx` file might be in your `src` folder if it isn't in your root folder.
    </Callout>

    ```tsx filename="/src/app/layout.tsx"

>     import { ClerkProvider } from '@clerk/nextjs'

Reviewer 2: ```suggestion
import { ClerkProvider } from '@clerk/nextjs'
import type { Metadata } from 'next'
import { Inter } from 'next/font/google'
import './globals.css'

const inter = Inter({ subsets: ['latin'] })
```

Reviewer 2: I feel pretty strongly we should keep at least the import for globals.css. If people using the quickstart use the 'copy' button then that disables their Tailwind/CSS and they may not realize it. 

I think there is value in importing Inter, but also fine with not having it. The new Quickstart repo should be adjusted to match the final version of this.

>     export const metadata = {

Reviewer 2: ```suggestion
    export const metadata: Metadata = {

```

      title: 'Next.js 13 with Clerk',
    }

    export default function RootLayout({
      children,
    }: {
      children: React.ReactNode
    }) {
      return (
        <ClerkProvider>
          <html lang="en">

>             <body>{children}</body>

Reviewer 2: ```suggestion
            <body className={inter.className}>{children}</body>

```

          </html>
        </ClerkProvider>
      )
    }
    ```

    <Callout type="warning">
      The root layout is a server component. If you plan to use the `<ClerkProvider />` outside the root layout, it will need to be a server component as well.
    </Callout>
  </Tab>
  <Tab>
    ```tsx filename="_app.tsx"

>       import { ClerkProvider } from "@clerk/nextjs";

Reviewer 2: ```suggestion
      import '@/styles/globals.css'
      import { ClerkProvider } from "@clerk/nextjs";

```

}
```

> ```tsx filename="pages/example.tsx" {1,7}

Reviewer 2: ```suggestion
```tsx filename="pages/index.tsx" {1,7}
```

import { UserButton } from "@clerk/nextjs";

> export default function Example() {

Reviewer 2: ```suggestion
export default function Home() {
```

</Steps>

## You're authenticated!

Congratulations! Your app is now only available to authenticated users! But this is just the first step. 

### Get the finished project

<Tabs items={["App Router version", "Page Router version"]}>
  <Tab>
    [`github.com/clerkinc/clerk-next-app-router-starter`](https://github.com/clerkinc/clerk-next-app-router-starter)
  </Tab>
  <Tab>
    [`github.com/clerkinc/clerk-nextjs-starter`](https://github.com/clerkinc/clerk-nextjs-starter)
  </Tab>
</Tabs>

# Next steps

<div className="container mx-auto my-4">
  <div className="grid grid-cols-1 gap-6 md:grid-cols-2">
    <Cards title="Create custom sign-up and sign-in pages" description="Learn how add custom sign-up and sign-in pages with Clerk components." link="/docs/references/nextjs/custom-signup-signin-pages" cta="Learn More" />

    <Cards title="Read user and session data" description="Learn how to use Clerk's to access the active session and user data in your Next.js application." link="/docs/references/nextjs/read-sessions-data" cta="Learn More" />

>     <Cards title="Client Side Helpers" description="Learn more about our client side helpers and how to use them." link="/docs/references/nextjs/overview#client-side-helpers" cta="Learn More" />

Reviewer: ```suggestion
    <Cards title="Client Side Helpers" description="Learn more about Next.js client side helpers and how to use them." link="/docs/references/nextjs/overview#client-side-helpers" cta="Learn More" />

```
